File: android/src/main/kotlin/com/example/easy_video_editor/handler/CompressVideoCommand.kt
package com.example.easy_video_editor.handler

import android.content.Context
import androidx.media3.common.util.UnstableApi
import com.example.easy_video_editor.command.Command
import com.example.easy_video_editor.utils.VideoUtils
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.cancel
import kotlinx.coroutines.launch

class CompressVideoCommand(private val context: Context) : Command {
    @UnstableApi
    override fun execute(call: MethodCall, result: MethodChannel.Result) {
        val videoPath = call.argument<String>("videoPath")
        val targetHeight = call.argument<Number>("targetHeight")?.toInt() ?: 720 // Default to 720p

        if (videoPath == null) {
            result.error(
                "INVALID_ARGUMENTS",
                "Missing required argument: videoPath",
                null
            )
            return
        }

        // Create a new scope that's tied only to this method call
        val methodScope = CoroutineScope(Dispatchers.Main + Job())
        
        methodScope.launch {
            try {
                val outputPath = VideoUtils.compressVideo(
                    context = context,
                    videoPath = videoPath,
                    targetHeight = targetHeight
                )
                result.success(outputPath)
            } catch (e: Exception) {
                result.error("COMPRESS_ERROR", e.message, null)
            } finally {
                methodScope.cancel()
            }
        }
    }
}

File: ios/Classes/handler/CompressVideoCommand.swift
import Flutter
import AVFoundation

class CompressVideoCommand: Command {
    func execute(call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let arguments = call.arguments as? [String: Any],
              let videoPath = arguments["videoPath"] as? String,
              let targetHeight = arguments["targetHeight"] as? Int else {
            result(FlutterError(
                code: "INVALID_ARGUMENTS",
                message: "Missing required arguments: videoPath or targetHeight",
                details: nil
            ))
            return
        }
        
        DispatchQueue.global(qos: .userInitiated).async {
            do {
                let outputPath = try VideoUtils.compressVideo(
                    videoPath: videoPath,
                    targetHeight: targetHeight
                )
                
                DispatchQueue.main.async {
                    result(outputPath)
                }
            } catch VideoError.fileNotFound {
                DispatchQueue.main.async {
                    result(FlutterError(
                        code: "FILE_NOT_FOUND",
                        message: "The video file was not found at the specified path",
                        details: nil
                    ))
                }
            } catch VideoError.invalidAsset {
                DispatchQueue.main.async {
                    result(FlutterError(
                        code: "INVALID_ASSET",
                        message: "Could not create video asset from the provided path",
                        details: nil
                    ))
                }
            } catch VideoError.exportFailed(let message) {
                DispatchQueue.main.async {
                    result(FlutterError(
                        code: "EXPORT_FAILED",
                        message: message,
                        details: nil
                    ))
                }
            } catch {
                DispatchQueue.main.async {
                    result(FlutterError(
                        code: "COMPRESS_ERROR",
                        message: error.localizedDescription,
                        details: nil
                    ))
                }
            }
        }
    }
}
